### ISSUE TITLE
Performance stuff again (Correlation by Occupancy Type)

### ISSUE THREAD

**Issue by AlexRyanUSACE** [HydrologicEngineeringCenter/HEC-FDA#748]
Simulations run very slow for even medium sized inventories (test simulation with ~250 RES structures takes 16 minutes to compute, with max of 5000 iterations and 0.05 convergence tolerance). It appears this is just down to the sheer number of calculations performed for each structure. In the example above the inverseCDF function is called well over 12 billion times, and eats up the majority of cpu cycles. 

Some half-formed thoughts on this; one idea would be, since it's running billions of functions to sample random values, it would probably be significantly faster to pre-load a set number of random samples (well under a billion would probably be sufficient for nearly all cases) and then randomly draw from that sample instead of computing the inverse cdf. This could be done at multiple possible different levels, and testing would need to be done to identify which would be the best (if indeed there would be one optimal level); distribution samples would be the most obvious, but it may end up being more performant to sample full occtype permutations or something along those lines. It may also be necessary if doing this to add a separate function to compute the optimal sample size for these preloaded pools. This would essentially substitute upfront fixed costs for variable costs on cpu time that would scale with iterations, structures, WSEs, etc.

Another idea would be to sample structure uncertainty at the occtype level instead of the structure level, which would dramatically speed up (and equalize) runtimes. This would (probably?) come at the cost of better capturing uncertainty by not artificially correlating uncertainty, but the resulting mean outputs should be the same and uncertainty bounds higher. From a purely theoretical standpoint there is probably an argument to be made that the occtype correlation should be 0 < correlation < 1, so these two options are probably on the extreme of what would be ideal. If nothing else, if correlation of 0 involves significantly longer runtimes, the ui should maybe contain an option to correlate occtypes.

**Comment by HenryGeorgist** [comment]
another way of thinking about the uncertainty (instead of correlated vs not correlated) is to think of it as a natural variability or a knowledge uncertainty. perfect correlation aligns with viewing it as a knowledge uncertainty (we dont know what the damage is for a structure of a class for a given depth, more data would reduce our uncertainty). perfectly uncorrelated aligns with it being a natural variability (the uncertainty of damage for a given depth naturally varies amongst structures within an occupancy type - and it cannot be reduced with more information)

**Comment by AlexRyanUSACE** [comment]
> 

Yeah, and i think it's a bit of both right? Which if so is why, in the terms of correlation on the computation side, i would think ideally you would want it somewhat correlated; to some extent we are sampling realizations from our uncertainty about depth damage etc for a given occupancy type, but also attempting to account for natural variability between structures of that occtype. So in a given iteration we would be making an assumption about both structure type level relationships as well as individual structure level variance. 

I don't think it's worth, if that truly is the ideal, actually pursuing some partial correlation approach, as the only practical impact of this assumption that i know of is uncertainty bounds (i guess damage exceedance probability also?), but i guess what i was thinkin is that if the true answer is somewhere in the middle, we're kind of choosing between two poles two find the one that minimizes error.

**Comment by AlexRyanUSACE** [comment]
I forgot to mention in my initial comment that i swapped out all the distributions in my test model for uniforms to replace the normal inverse cdf with a very simple math operation and it took the compute time for the first damage category from 16 minutes down to 10. This is probably the upper bound of time savings, as i would guess randomly pulling a value from a list is probably more costly than basic arithmetic, but i don't know enough about that stuff to say for sure.

**Comment by Brennan1994** [comment]
Closing discussion as stale. Occ type sampling has been long since reworked.
